File: XIV/Delphi/Database/IntraWeb_DataSnap_Biolife/IntraWebClient/ServerController.pas
unit ServerController;
interface
uses
  SysUtils, Classes, IWServerControllerBase, IWBaseForm, HTTPApp,
  // For OnNewSession Event
  UserSessionUnit, IWApplication, IWAppForm, DataModuleUnit, IWDataModulePool;

type
  TIWServerController = class(TIWServerControllerBase)
    Pool: TIWDataModulePool;
    procedure IWServerControllerBaseCreate(Sender: TObject);

    procedure PoolCreateDataModule(var ADataModule: TDataModule);
    procedure PoolFreeDataModule(var ADataModule: TDataModule);
    procedure IWServerControllerBaseNewSession(aSession: TIWApplication);
  private

  public
  end;

  function UserSession: TIWUserSession;
  function IWServerController: TIWServerController;
  function LockDataModule: TDataModule1;
  procedure UnlockDataModule(ADataModule: TDataModule1);

implementation

{$R *.dfm}

uses
 IWInit, IWGlobal;

function UserSession: TIWUserSession;
begin
  Result := TIWUserSession(WebApplication.Data);
end;

function IWServerController: TIWServerController;
begin
  Result := TIWServerController(GServerController);
end;

procedure TIWServerController.IWServerControllerBaseCreate(Sender: TObject);
begin
  Pool.Active := True;
  //IWServerController.SessionTrackingMethod:=tmURL;
end;

procedure TIWServerController.IWServerControllerBaseNewSession(
  aSession: TIWApplication);
begin
  ASession.Data := TIWUserSession.Create(nil, ASession );
end;

procedure TIWServerController.PoolCreateDataModule(var ADataModule: TDataModule);
begin
  ADataModule := TDataModule1.Create(nil);
end;

procedure TIWServerController.PoolFreeDataModule(var ADataModule: TDataModule);
begin
  FreeAndNil(ADataModule);
end;

function LockDataModule: TDataModule1;
begin
  Result := TDataModule1(TIWServerController(GServerController).Pool.Lock);
end;

procedure UnlockDataModule(ADataModule: TDataModule1);
var
  LTemp: TDataModule;
begin
  LTemp := ADataModule;
  TIWServerController(GServerController).Pool.Unlock(LTemp);
end;


initialization
  TIWServerController.SetServerControllerClass;
end.

File: 15/C++/DemoCPP_FastMM4/ServerController.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "ServerController.h"
#include "IWGlobal.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma resource "*.dfm"

//---------------------------------------------------------------------------
__fastcall TIWServerController::TIWServerController(TComponent* Owner)
        : TIWServerControllerBase(Owner)
{
}
//---------------------------------------------------------------------------


void __fastcall TIWServerController::IWServerControllerBaseNewSession(
      TIWApplication *ASession)
{
  ASession->Data = new TIWUserSession(NULL, ASession);
}
//---------------------------------------------------------------------------


//---------------------------------------------------------------------------

void __fastcall TIWServerController::IWServerControllerBaseCreate(
      TObject *Sender)
{
   Pool->Active = true;
}
//---------------------------------------------------------------------------
void __fastcall TIWServerController::PoolCreateDataModule(
      TDataModule *&ADataModule)
{
  ADataModule = new TDataModule1(NULL);
}
//---------------------------------------------------------------------------
void __fastcall TIWServerController::PoolFreeDataModule(
      TDataModule *&ADataModule)
{
  ADataModule->Free();
}
//---------------------------------------------------------------------------

TDataModule1* LockDataModule()
{
  return (TDataModule1*)((TIWServerController*)gServerController)->Pool->Lock();
}

//---------------------------------------------------------------------------

void UnlockDataModule(TDataModule1* ADataModule)
{
  TDataModule* temp = ADataModule;
  ((TIWServerController*)gServerController)->Pool->Unlock(temp);
}
//---------------------------------------------------------------------------
void setServerController() {
  TIWServerController::SetServerControllerClass();
}
//---------------------------------------------------------------------------

TIWServerController* IWServerController()
{
  return (TIWServerController*)gServerController;
}

#pragma startup setServerController
